File: solutions/java/need-for-speed/1/src/main/java/NeedForSpeed.java
class NeedForSpeed {
    private int distance;
    private int battery;
    private int speed;
    private int batteryDrain;

    public NeedForSpeed(int speed, int batteryDrain) {
        this.battery = 100;
        this.speed = speed;
        this.batteryDrain = batteryDrain;
    }

    public boolean batteryDrained() {
        return battery == 0;
    }

    public int distanceDriven() {
        return distance;
    }

    public void drive() {
        if (battery - batteryDrain < 0) {
            return;
        }

        battery -= batteryDrain;
        distance += speed;
    }

    public static NeedForSpeed nitro() {
        return new NeedForSpeed(50, 4);
    }
}

class RaceTrack {
    private final int distance;

    public RaceTrack(int distance) {
        this.distance = distance;
    }

    public boolean tryFinishTrack(NeedForSpeed car) {
        
        while (!car.batteryDrained()) {
            car.drive();
        }

        return car.distanceDriven() >= distance;
    }
}

File: solutions/csharp/need-for-speed/1/NeedForSpeed.cs
using System;
class RemoteControlCar
{
    private int speed;
    private int batteryDrain;
    private int distance;
    private int battery = 100; 
    public RemoteControlCar(int speed, int BatteryDrain)
    {
        this.speed = speed;
        this.batteryDrain = BatteryDrain;
    }
    public bool BatteryDrained()
    {
        return battery < batteryDrain;
    }
    public int DistanceDriven()
    {
        return distance; 
    }
    public void Drive()
    {
        if (!BatteryDrained())
        {
            this.distance += speed;
            this.battery -= this.batteryDrain;
            Console.WriteLine(this.battery);
        }
    }
    public static RemoteControlCar Nitro()
    {
        return new RemoteControlCar(50, 4);
    }
}
